objc/objc_runtime_darwin.go

```diff
@@ -148,203 +170,165 @@ func AllocateClassPair(super Class, name string, extraBytes uintptr) Class {
+type IvarAttrib int
 
-// MismatchError occurs when the Go struct definition doesn't match that of Objective-C
-var MismatchError = errors.New("go struct doesn't match objective-c struct")
+const (
+	ReadOnly IvarAttrib = 1 << iota
+	ReadWrite
+)
 
-// RegisterClass takes a pointer to a struct that implements the Selector interface.
-// It will register the structs fields and pointer receiver methods in the Objective-C
-// runtime using the SEL returned from Selector. Any errors that occur trying to add
-// a Method or Ivar is returned as an error. Such errors may occur in parsing or because
-// the size of the struct does not match the size in Objective-C. If no errors occur
-// then the returned Class has been registered successfully.
-//
-// The struct's first field must be of type Class and have a tag that matches the format
-// `objc:"ClassName : SuperClassName <Protocol, ...>`. This tag is equal to how the class
-// would be defined in Objective-C.
-func RegisterClass(object Selector) (Class, error) {
-	ptr := reflect.TypeOf(object)
-	strct := ptr.Elem()
-	if strct.NumField() == 0 || strct.Field(0).Type != reflect.TypeOf(Class(0)) {
-		return 0, fmt.Errorf("objc: need objc.Class as first field: %w", MismatchError)
-	}
-	isa := strct.Field(0)
-	tag := isa.Tag.Get("objc")
-	if tag == "" {
-		return 0, fmt.Errorf("objc: missing objc tag: %w", TagFormatError)
-	}
-	// split contains the class name and super class name followed by all the Protocols
-	// start with two for ClassName : SuperClassName
-	var split = make([]string, 2)
-	{
-		// This is a simple parser for the objc tag that looks for the format
-		//  	"ClassName : SuperClassName <Protocol, ...>"
-		// It appends to the split variable with the [ClassName, SuperClassName, Protocol, ...]
-
-		var i int  // from tag[0:i] is whatever identifier is next
-		var r rune // r is the current rune
-		skipSpace := func() {
-			for _, c := range tag {
-				if c == ' ' {
-					tag = tag[1:]
-					continue
-				}
-				break
-			}
-		}
-		skipSpace()
-		// get ClassName
-		for i, r = range tag {
-			if r == ' ' || r == ':' {
-				break
-			}
-		}
-		split[0] = tag[0:i] // store ClassName
-		tag = tag[i:]
+// FieldDef is a definition of a field to add to an Objective-C class.
+// The name of the field is what will be used to access it through the Ivar. If the type is bool
+// the name cannot start with `is` since a getter will be generated with the name `isBoolName`.
+// The name also cannot contain any spaces.
+// The type is the Go equivalent type of the Ivar.
+// Attribute determines if a getter and or setter method is generated for this field.
+type FieldDef struct {
+	Name      string
+	Type      reflect.Type
+	Attribute IvarAttrib
+}
 
-		skipSpace()
+// ivarRegex checks to make sure the Ivar is correctly formatted
+var ivarRegex = regexp.MustCompile("[a-z][a-zA-Z0-9]*")
```

Comment:
I realized a name can incude `_`. So

`[a-z_][a-zA-Z0-9_]`

is more exact?

https://google.github.io/styleguide/objcguide.html#category-naming